connect-inject/container_init.go

```diff
@@ -72,11 +72,19 @@ func (h *Handler) containerInit(pod corev1.Pod, k8sNamespace string) (corev1.Con
 	}
 
 	// When ACLs are enabled, the ACL token returned from `consul login` is only
-	// valid for a service with the same name as the ServiceAccountName.
-	if data.AuthMethod != "" && data.ServiceName != pod.Spec.ServiceAccountName {
+	// valid for a service with the same name as the ServiceAccountName. If the
+	// annotation for service name is set, check against that.
+	if data.AuthMethod != "" && data.ServiceName != "" && data.ServiceName != pod.Spec.ServiceAccountName {
 		return corev1.Container{}, fmt.Errorf("serviceAccountName %q does not match service name %q", pod.Spec.ServiceAccountName, data.ServiceName)
 	}
 
+	// If the service name annotation does not exist, set the service account name flag contents for the init container
+	// command. If the init container sees the service account name flag as non-empty, it will check it against the
+	// Kubernetes service name to ensure they are the same.
+	if data.ServiceName == "" {
```

Comment:
Minor nit : I think everything will "just work" but if the user has a serviceaccount and ACLs are disabled in consul, we're filling this in for no reason. It wont end up templated into the command so it should work.
But should we guard it with `if data.AuthMethod != "" && ` for readability sake?

Reply:
That's a great point, I'll add that!